mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -438,11 +438,23 @@ class Word2Vec extends Serializable with Logging {
             None
           }
         }.flatten
-      }
-      val synAgg = partial.reduceByKey { case (v1, v2) =>
-          blas.saxpy(vectorSize, 1.0f, v2, 1, v1, 1)
-          v1
+      }.persist()
+      // SPARK-24666: do normalization for aggregating weights from partitions.
+      // Original Word2Vec either single-thread or multi-thread which do Hogwild-style aggregation.
+      // Our approach needs to do extra normalization, otherwise adding weights continuously may
+      // cause overflow on float and lead to infinity/-infinity weights.
+      val keyCounts = partial.countByKey()
+      val synAgg = partial.mapPartitions { iter =>
+        iter.map { case (id, vec) =>
+          val v1 = Array.fill[Float](vectorSize)(0.0f)
+          blas.saxpy(vectorSize, 1.0f / keyCounts(id), vec, 1, v1, 1)
+          (id, v1)
+        }
+      }.reduceByKey { case (v1, v2) =>
```

Comment:
I can only do averaging like this. The group key can not be accessed in `reduceByKey`.

Reply:
What if you emit `(id, v1, 1)` above and then sum those 1s as a count, and then divide through after `reduceByKey`? I think it's _possible_, just not 100% sure it's the right thing to do. But sounds quite plausible.

Reply:
During `reduceByKey`, we already do sum up and can lead to infinity? Once it is done, i think it does not make sense anymore to divide?

Reply:
oh, I got your idea. If it is not suddenly overflow, but incrementally, then we can divide it after `reduceByKey`.

Reply:
I haven't though it through, but yeah if you can normalize between iterations it should be OK. otherwise yeah you have to compute some kind of running average, which isn't super hard.

Reply:
yeah, I did what you suggested above by emitting many 1s and do average after `reduceByKey`.